Query editor path intellisense

tbui17 opened the pull request:
fixes #140 

# Context

I used VSCode's Typescript path intellisense with imports as inspiration for required behavior. In addition to what it currently supports, intellisense will also pick up `\`.

On each keypress, the editor will read the current line and check if it looks like an IO command (i.e. `.save` or `.load`) and it looks like a file path is being typed for an argument. If so, we try a series of retrieval strategies based on the pattern of the current path and find the first one that works.

The completion window has its own internal filtering algorithm dependent on its own state at the time of opening, so I can't fully separate intellisense out into a separate unit or else I'd introduce some inconsistencies in filtering behavior as the user types or erases. For now, the service defers filtering to the client based on provided data. 
 
# Changes
- Created new intellisense project with file system intellisense service and test project.
- Added `TestableIO.System.IO.Abstractions` and related packages for file system mocking
  - Kept dependencies to a minimum otherwise, didn't want to change too much for this initial PR.
- Moved `IntellisenseEntry` into intellisense project
- Added IO command parser
- Changed some ordering of completion window setup logic to support integrating intellisense

NeilMacMullen commented:
This is *really* great - thank you!

tbui17 commented:
> This is _really_ great - thank you!

Happy to help! Cleaning up the code a bit before reviews.

NeilMacMullen commented:
No worries- I took a look at the first PR and it seemed pretty good (already useful enough that I was using a local lokqldx build for some projects!)

btw - without at all wanting to put you under any obligations to extend the scope further ;-) you are welcome to change the current popup behaviour further to make it more compatible with VSCode  behaviour!

tbui17 commented on libraries/Intellisense/FileSystem/RootedPath.cs:
```diff
@@ -0,0 +1,50 @@
+/// </summary>
+internal readonly record struct RootedPath
+{
+    public string Value { get; }
```
Since `Path` methods work on strings, I didn't think it was worth it to encapsulate the state.

tbui17 commented on applications/lokqlDx/QueryEditor.xaml.cs:
```diff
@@ -444,3 +497,26 @@ public void SetSchema(SchemaLine[] schema)
+            return null;
+        }
+
+        var isIoCommand = args[0].Equals(".save", StringComparison.OrdinalIgnoreCase) ||
+                          args[0].Equals(".load", StringComparison.OrdinalIgnoreCase);
```
I could probably fetch all of the supported commands from `CommandProcessor.GetVerbs` in a different PR.

NeilMacMullen replied:
Yes, what I had in mind (eventually) was to  introduce some registration mechanism for commands so that they could advertise that (for example) they could accept a file path

tbui17 commented:
I'd be happy to take on intellisense as a whole, including functionality, testing, and architecture. I want to make sure my general approach was fine with this initial PR, though.

NeilMacMullen commented:
:-) sounds good.  Been a bit busy with family commitments last few days but will take a proper look at this tomorrow.

tbui17 commented:
No rush! Please, take your time.

NeilMacMullen commented on libraries/Intellisense/FileSystem/FileSystemIntellisenseService.cs:
```diff
@@ -0,0 +1,43 @@
+            new SiblingRootedPathCompletionResultRetriever(reader)
+        ];
+
+        _retrievers =
+        [
+            new RootedPathCompletionResultRetriever(rootedPathRetrievers)
+        ];
```
nice !  Bear in mind the concept of the "current working directory" is a bit loose atm.  By default, once a workspace has been saved, the kusto.datapath setting in SettingsProvider is set to the folder containing the workspace and any unrooted path is expected to be relative to that, _but_ the user can explicitly set kusto.datapath themselves as part of the workspace startup script.

NeilMacMullen commented on libraries/Intellisense/FileSystem/FileSystemIntellisenseService.cs:
```diff
@@ -0,0 +1,43 @@
+        }
+        catch (IOException e)
+        {
+            Console.Error.WriteLine(e);
```
Not sure if you want to keep this for debugging?

tbui17 replied:
Looking back at it now, it really doesn't make sense for me to leave it in there as is. I was intending to have a placeholder   for a logger like `Microsoft.Extensions.Logging`. The package also provides the interface, so I didn't want to do needless work either.

I suppose a `//TODO` would have been better here.

NeilMacMullen replied:
Yeah it's a tricky one - there is a mechanism for reporting stuff to the user via the current context (exp.Warn) but I don't think we'd want to worry them about this kind of thing - has more of the look of "ugh we can't current see a bit of the filesystem, tough luck"

NeilMacMullen commented on libraries/Intellisense/FileSystem/RootedPath.cs:
```diff
@@ -0,0 +1,50 @@
+        if (!Path.IsPathRooted(path))
+        {
+            return null;
+        }
```
Genuine question is there a case where we ever try to create a RootedPath from a non-rooted string?  I'm generally a proponent of avoiding propagating nulls as legitimate values (I do like static constructor/factories though!)

NeilMacMullen replied:
I haven't traced the full logic so possibly I'm missing something but if I'm not mistaken, then `GetPathIntellisenseOptions` uses `CompletionResult.Empty` to signal there are no completions which means the test it uses in FirstOrDefault could simply be elided....
```
return _retrievers
    .Select(x => x.GetCompletionResult(path))
    .Append(CompletionResult.Empty)
    .FIrst();
```
if GetCompletionResult returns CompletionResult.Empty where it would otherwise return null

That would require RootedPath.Create etc to return some concept of a RootedPath.Empty (is `RootedPath(string.Empty)` a "null object in that sense?)

NeilMacMullen replied:
Just got an email notification of a comment you made (and maybe deleted?)  Yeah sentence above about "avoiding propagating nulls" was a bit convoluted.  So to be clear.. I hate nulls ;-) and go to great lengths to avoid them turning up as values in code.   (Clearly that is not always possible when interfacing with other code).  The [null object pattern](https://en.wikipedia.org/wiki/Null_object_pattern) is my preferred approach.

tbui17 replied:
Same with `RootedPath` and `ParentChildPathPair`. Will see if I can make the API consistent.

NeilMacMullen commented on libraries/Intellisense/FileSystem/RootedPath.cs:
```diff
@@ -0,0 +1,50 @@
+    public bool IsRootDirectory()
+    {
+
+        if (Value[0].IsDirectorySeparator())
```
another corner-case to consider (not required in this PR unless you are particularly keen) is that UNC paths can start with double-slashes (in either direction!)...
```
ls //cinderella/c
ls \\cinderella\c
```

NeilMacMullen replied:
👍 It's frustrating that while the file API is quite happy to use UNC paths, there doesn't seem to be very good support for detecting/parsing them in the Path api.... best approach I could find elsewhere was to write my own IsRooted variant that checked for double slash

NeilMacMullen commented on libraries/Intellisense/PathExtensions.cs:
```diff
@@ -0,0 +1,63 @@
+        return path == Path.DirectorySeparatorChar || path == Path.AltDirectorySeparatorChar;
+    }
+
+    public static string? GetNonEmptyParentDirectory(this string path)
```
per comments about nullability, is it possible to use string.Empty as the "null" value here ?  (the name suggests not...)   FWIW there are various extension methods in the NotNullStrings extension library I use eg... 

`someStringVar.NullToEmpty()`    `somePotentiallNullString.OrWhenBlank("other value")`

tbui17 replied:
I think the logic for this could probably be inlined, along with an explicit unreachable exception thrown if the path somehow ended up being empty or whitespace within the context of a rooted path.

For reference:
https://learn.microsoft.com/en-us/dotnet/api/System.IO.Path.GetDirectoryName?view=net-9.0

NeilMacMullen commented:
@tbui17 FYI I'm planning to make a new release in the next few days.  Happy to merge this "as is" since it's already super useful but will leave open for tonight.

tbui17 commented:
@NeilMacMullen I'll get around to wrapping this up in a bit.

tbui17 commented:
# Changes
- Added handling for UNC paths
- Removed console log
- Moved path related classes into own folder
- Changed retrievers to work with an `IFileSystemPath`, which could be an `EmptyPath`. Interactions with it have been modified accordingly to reduce total amount of branching conditions. Names were changed and abstracted code was inlined to be compatible with the new semantics.

NeilMacMullen commented:
Sounds good. Away today but will see if I can figure out the right git runes tomorrowOn 22 Mar 2025 15:41, tbui17 ***@***.***> wrote:
  Quite frankly, I think it's a little difficult to review how I did the refactoring from the diff. Perhaps it would be better to just merge from prior to the changes f31f53b, then I can split up these changes and have a proper review of the code myself as well. I'd also want to get CI up and going for proper testing of file shares.—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>



tbui17 left a comment (NeilMacMullen/kusto-loco#147)
Quite frankly, I think it's a little difficult to review how I did the refactoring from the diff. Perhaps it would be better to just merge from prior to the changes f31f53b, then I can split up these changes and have a proper review of the code myself as well. I'd also want to get CI up and going for proper testing of file shares.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

tbui17 commented:
Thanks. If you're running short on time, I can revert the branch back to that point and force push.

NeilMacMullen commented:
If you're able to do that, it would be great - also less chance of me getting it wrong 🙂
________________________________
From: tbui17 ***@***.***>
Sent: 22 March 2025 17:07
To: NeilMacMullen/kusto-loco ***@***.***>
Cc: NeilMacMullen ***@***.***>; Mention ***@***.***>
Subject: Re: [NeilMacMullen/kusto-loco] Query editor path intellisense (PR #147)


Thanks. If you're running short on time, I can revert the branch back to that point and force push.

—
Reply to this email directly, view it on GitHub<https://github.com/NeilMacMullen/kusto-loco/pull/147#issuecomment-2745369423>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ACFVKSLOBMY4FT7L3G7Z2R32VWKETAVCNFSM6AAAAABYSFUCJKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONBVGM3DSNBSGM>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

[tbui17]tbui17 left a comment (NeilMacMullen/kusto-loco#147)<https://github.com/NeilMacMullen/kusto-loco/pull/147#issuecomment-2745369423>

Thanks. If you're running short on time, I can revert the branch back to that point and force push.

—
Reply to this email directly, view it on GitHub<https://github.com/NeilMacMullen/kusto-loco/pull/147#issuecomment-2745369423>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ACFVKSLOBMY4FT7L3G7Z2R32VWKETAVCNFSM6AAAAABYSFUCJKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONBVGM3DSNBSGM>.
You are receiving this because you were mentioned.Message ID: ***@***.***>
